ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -0,0 +1,140 @@
+  private final QuotaSource quotaSource;
+  private final float maxFrontendCuUsageToAllowExceed;
+  private final long throttleRetryAfterMs;
+
+  /**
+   * Constructor for {@link AmbryCapacityUnitQuotaEnforcer}.
+   * @param quotaSource {@link QuotaSource} where the quota limit and usage will be saved and retrieved from.
+   */
+  public AmbryCUQuotaEnforcer(QuotaSource quotaSource, QuotaConfig quotaConfig) {
+    this.quotaSource = quotaSource;
+    this.throttleRetryAfterMs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    this.maxFrontendCuUsageToAllowExceed = quotaConfig.maxFrontendCuUsageToAllowExceed;
+  }
+
+  @Override
+  public void init() {
+  }
+
+  @Override
+  public void charge(RestRequest restRequest, Map<QuotaName, Double> requestCostMap) throws QuotaException {
+    final QuotaName quotaName = QuotaUtils.getCUQuotaName(restRequest);
+    if (requestCostMap.isEmpty() || !requestCostMap.containsKey(quotaName)) {
+      String errorMessage = String.format("No %s cost provided for request %s. Nothing to charge", quotaName.name(),
+          RestUtils.convertToStr(restRequest));
+      throw new QuotaException(errorMessage, true);
+    }
+    final QuotaResource quotaResource = QuotaResource.fromRestRequest(restRequest);
+
+    doAndHandleException(() -> {
+      quotaSource.chargeUsage(quotaResource, quotaName, requestCostMap.get(quotaName));
+      return null;
+    }, String.format("Could not charge for request %s due to", restRequest));
+  }
+
+  @Override
+  public QuotaRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    final QuotaName quotaName = QuotaUtils.getCUQuotaName(restRequest);
+    float usage = doAndHandleException(() -> {
+      final QuotaResource quotaResource = QuotaResource.fromRestRequest(restRequest);
+      return quotaSource.getUsage(quotaResource, quotaName);
+    }, String.format("Could not recommend for request %s due to", restRequest));
+    return buildQuotaRecommendation(usage, quotaName);
+  }
+
+  @Override
+  public boolean isQuotaExceedAllowed(RestRequest restRequest) throws QuotaException {
+    return doAndHandleException(() -> quotaSource.getSystemResourceUsage(QuotaUtils.getCUQuotaName(restRequest))
+            < maxFrontendCuUsageToAllowExceed,
+        String.format("Could not check quota exceed allowed for request %s due to", restRequest));
+  }
+
+  @Override
+  public List<QuotaName> supportedQuotaNames() {
+    return SUPPORTED_QUOTA_NAMES;
+  }
+
+  @Override
+  public QuotaSource getQuotaSource() {
+    return quotaSource;
+  }
+
+  @Override
+  public void shutdown() {
+
+  }
+
+  /**
+   * Build the {@link QuotaRecommendation} object from the specified usage and {@link QuotaName}.
+   * @param usage percentage usage.
+   * @param quotaName {@link QuotaName} object.
+   * @return QuotaRecommendation object.
+   */
+  private QuotaRecommendation buildQuotaRecommendation(float usage, QuotaName quotaName) {
+    boolean shouldThrottle = (usage >= 100);
```

Comment:
I think we should always recommend throttling only when customer has used up their quota. I don't see any use case where we would want to throttle someone before they used up quota.
We should definitely warn customers that their quota is going to be over soon. We have logic in `SimpleQuotaRecommendationMergePolicy` to generate that warning, and that will work irrespective of how we recommend `shouldThrottle` here.

I do agree with you that it will be better to add configs to define the thresholds at which we issue warning. Currently it is hardcoded in `SimpleQuotaRecommendationMergePolicy`. I will push out an RB to make those values configurable.